### DISCUSSION TITLE
How to manually create a window and manipulate it

### DISCUSSION THREAD

**Discussion by fralonra** [tauri-apps/tauri#2283] [Q&A] [answered]
Hi all, I'm using Tauri (beta version) to build a project.

I'm wondering how to create a window from Rust backend and manipulate the window afterward.

I have seen methods like `create_window`, `show`, `hide` etc, but have no idea how to use them for there is little doc about this topic.

Thanks for reply.

**Comment by amrbashir** [comment, marked as answer]
you have to get an app handle or window handle which you can obtain in any hook on the tauri builder. 

note: code below doesn't have any arguments just for simplicity, check the urls to see the arguments needed.
```rs
tauri::Builder::default()
	.setup(|app|{
		let window = app.create_window(); // https://tauri.studio/en/docs/api/rust/tauri/window/struct.Window#create_window
	})
	.on_page_load(|window, _| {
		let window = window.create_window();
	})
	.run();
``` 
or from js you can just use the `WebviewWindow` constructor
```js
import { WebviewWindow } from "@tauri-apps/api/window";
// or `withGlobalTauri: true` in tauri.conf.json
const { WebviewWindow } = window.__TAURI__.window;

new WebviewWindow() // https://tauri.studio/en/docs/api/js/classes/window.webviewwindow/

```

**Reply by fralonra** [reply]
Thanks for the reply.

I'm now able to access the App in setup hook.

But there is still some questions. I want to create multiple windows and each of them has different configuration. Below lies my code example but I don't think it is the correct way to do it:

```rust
builder
      .setup(|app| {
            // create multiple windows.
            app.create_window();
            // ...
      })
      .on_page_load(window, _| {
            // customize windows here:
            window.set_title();
            window_set_size();
            // ...
      })
      .run(tauri::generate_context!())
```

The above code has some defects like: customizing the window in `on_page_load` hook instead of at the int time of the window; can't set transparency for there is no such method on `tauri::Window`.

Can anyone help me with this?

**Reply by amrbashir** [reply]
the third paramter of `create_window` is a closure where you get access to `WindowAttributes` builder that you can mutate it in the closure then return it again. 

Can't give you a full example now as I am on my phone, join our discord community and we will guide you through.

**Reply by fralonra** [reply]
> the third paramter of `create_window` is a closure where you get access to `WindowAttributes` builder that you can mutate it in the closure then return it again.
> 
> Can't give you a full example now as I am on my phone, join our discord community and we will guide you through.

Sorry I can't use discord on weekdays. Still I can't find any information about how to modify the webview window through WebviewAttributes.

**Reply by amrbashir** [reply]
here is a full example 
```rs
use tauri::WindowBuider;

tauri::Builder::default()
	.setup(|app|{
		app.create_window("window2".to_string(), WindowUrl::default(), |WinAttrs, WebviewAttrs| {
      		let WinAttrs = WinAttrs
        	.title("Main 2")
        	.resizable(true)
       		.transparent(false)
        	.decorations(true)
        	.always_on_top(false)
        	.inner_size(800.0, 600.0)
        	.min_inner_size(300.0, 150.0)
        	.fullscreen(false);
       		(WinAttrs, WebviewAttrs)
       	});
	})
	.run();
```

**Reply by fralonra** [reply]
Thanks for the detail example!

I've already been working on this way though there are still some errors, but I assume that's not due to Tauri.

I was confused because I had misunderstood that I should modify `WindowAttributes`, the second parameter of the closure.

Edited: things have been done, thanks!

**Reply by THEGOLDENPRO** [reply]
How do I do this in tauri 1.6.6, app doesn't have `.create_window()` for me. Was this removed?

**Comment by HuakunShen** [comment]
Is there a way to toggle a window's transparency?

I know `appWindow.setDecorations(false);` can be used to change decorations, but there is not an api for changing window transparency.

**Reply by FabianLars** [reply]
> Is there a way to toggle a window's transparency?

Not via the native apis, that can only be set on initialization. I guess you can just change the frontend's css to be opaque 🤔
